File: 246. Strobogrammatic Number.cpp
class Solution {
public:
    // two pointers: O(N), 9.27%, ok
    bool isStrobogrammatic(string num) {
        if(num.empty()) return true;
        unordered_map<char, char> dict = {{'6','9'}, {'0','0'}, {'1','1'}, {'8','8'}, {'9','6'}};
        int head = 0, tail = num.size() - 1;
        while(head <= tail){
            if(dict.count(num[head]) && dict[num[head]] == num[tail]){
                head++; tail--;
            }
            else return false;
        }
        return true;
    }

    // STL: O(N), 9.27%, ok
    bool isStrobogrammatic(string num) {
        if(num.empty()) return true;
        string test = "696 88 11 00";
        int head = 0, tail = num.size() - 1;
        string str1, str2;
        while(head <= tail){
            if(test.find(string(1, num[head]) + string(1, num[tail])) != string::npos){ // bug: num[head]+num[tail]
                head++; tail--;
            }
            else return false;
        }
        return true;
    }
};

File: 246. Strobogrammatic Number.py
class Solution(object):
    
    # hash table: O(N) time, O(1) space
    def isStrobogrammatic(self, num):
        """
        :type num: str
        :rtype: bool
        """
        pairs = {'6': '9', '9': '6', '1': '1', '0': '0', '8': '8'}
        for i in xrange((len(num)+1) / 2):
            if (num[i] not in pairs) or (pairs[num[i]] != num[len(num)-i-1]):
                return False
        return True
    
    # two pointer: O(N) time, O(1) space
    def isStrobogrammatic(self, num):
        """
        :type num: str
        :rtype: bool
        """
        pairs = ['69', '96', '11', '00', '88']
        for i in xrange((len(num)+1) / 2):
            if num[i] + num[len(num)-i-1] not in pairs:
                return False
        return True
            
        
        
        